src/strtab.rs

```diff
@@ -25,31 +26,31 @@ fn get_str(offset: usize, bytes: &[u8], delim: ctx::StrCtx) -> scroll::Result<&s
 }
 
 impl<'a> Strtab<'a> {
-    /// Construct a new strtab with `bytes` as the backing string table, using `delim` as the delimiter between entries
-    pub fn new(bytes: &'a [u8], delim: u8) -> Self {
-        Strtab {
+    /// Creates a `Strtab` directly without bounds check and without parsing it.
+    pub fn from_slice_unsafe(bytes: &'a [u8], offset: usize, len: usize, delim: u8) -> Self {
```

Comment:
e.g., this function here makes a lot of diff noise; also does this function need to be `pub`? one can construct a strtable now, but the vec isn't filled, which means any `get_at` calls would fail, which feels like the strtab is "half constructed"; also it is better to not introduce any new pub apis unless necessary, since once they're pub, we can't take them back

Reply:
> ok so other than `to_vec` being a breaking change, which if you feel strongly about, i guess i'm ok with it, the only other major concern i have is whether we're pessimizing 99% of binary's strtables in what seems like a very niche problem case.

Well, we have no data what the "usual" user of this crate is doing on average, so it is hard to optimize for.
But, as I said, this is only penalizing use cases where only a few symbols are requested once. All others such as:
* Symbols being requested multiple times
* Symbols having suffix overlap
* Most symbols being requested

Would not be hit at all or even profit from the pre-parsing.

> as far as i understand it, the only security implications I see are a kind of DoS on accessing strtables for these malformed binaries, but it's not clear to me how that is impactful, other than a tool taking a long time to run? it would be good to get some further motivations here.

I agree that many CLI tools some developer runs and cancels, if they take to long, are not affected by this.
However, think about the automated processes like CI services, etc. Furthermore, we run a decentralized network which validates ELF files on the fly, so it would easily grind to halt if somebody wanted it to, which is not acceptable.

> I'm not sure why, but I almost wish we had a variant constructor for Strtab that uses the pre-allocated form via a config we'd pass into parsing, and defaults to a non-preallocated form otherwise, but this is a much bigger PR. However, it might be worth doing, because a config for parsing options is something i've thought would be nice for a while.
> 
> with that in mind, i don't see anything in this PR that would prevent such a feature from landing in the future?
> 
> example:
> 
> ```rust
> enum StrtabKind {
>   Preallocated(Vec<&str>),
>   ZeroCopy
> }

Might not be so complicated, why not just default back to the original implementation of `get` in `get_at` if the `strings` `Vec` is empty, because it was constructed by `from_slice_unsafe` not by `parse`?